CT Body Sectioned Solid Horizontal

larswik opened the pull request:
Addresses  #225
So far only removed FixedAxisVertical and added CrossSectionPositions including rule ID.
I also started with the documentation. This will have to be reviewed and most probably updated further.

larswik committed:
CT Body Sectioned Solid Horizontal

Addresses  #225
So far only removed FixedAxisVertical and added CrossSectionPositions including rule ID.
I also started with the documentation. This will have to be reviewed and most probably updated further.

SergejMuhic commented on IFC4x3/Templates/Product Shape/Product Geometric Representation/Body Geometry/Body AdvancedSweptSolid Geometry/Body SectionedSolidHorizontal/Documentation.md:
```diff
@@ -0,0 +1,7 @@
+
+* _IfcShapeRepresentation_._RepresentationIdentifier_ = 'Body'
+* _IfcShapeRepresentation_._RepresentationType_ = 'AdvancedSweptSolid'
+* _IfcShapeRepresentation_._Items_ = _IfcSectionedSolidHorizontal
```
```suggestion
* _IfcShapeRepresentation_._Items_ = _IfcSectionedSolidHorizontal_
```

SergejMuhic reviewed (commented):
1. Think we need a ruleId on directrix because the template has no restrictions.
2. I do not think we need a RuleId on CrossSectionPositions as these are already restricted to `IfcAxis2PlacementLinear`.
3. RepresentationIdentifier = Body is currently missing

larswik committed:
Update IFC4x3/Templates/Product Shape/Product Geometric Representation/Body Geometry/Body AdvancedSweptSolid Geometry/Body SectionedSolidHorizontal/Documentation.md

Co-authored-by: SergejMuhic <sergejs1@gmail.com>

larswik committed:
Update DocTemplateDefinition.xml

larswik commented:
Removed RuleId for CrossSectionPositions and added the constraint RepresentationIdentifier='Body'. Regarding Directrix: Isn't there already a rule id?
I still get the uniqueIds "_000000000" that @jmirtsch fixed for CT sectioned Surface. Needs fixing before merging.

larswik committed:
Merge branch 'develop' into CT-Body-SectionedSolidHorizontal

larswik commented:
I think (hope) that I fixed the uniqueids when fixing merge conflicts. I added the constraint and then new rule ids were generated which resulted in a merge conflict. I just kept the old rule ids plus the constraint.

larswik commented:
Maybe misunderstood the comment about "Body missing" since there is already a constraint at "Body geometry" level.

larswik committed:
Update DocTemplateDefinition.xml

Remove the 'Body' constraint again.

SergejMuhic commented:
@larswik could you please fetch upstream? It seems your baseline is not the latest as `FixedAxisVertical` is still in `IfcSectionedSolidHorizontal`.

larswik commented:
@SergejMuhic which branch? I am up to date with upstream/develop.

AlexBrad1eyCT commented on IFC4x3/Templates/Product Shape/Product Geometric Representation/Body Geometry/Body AdvancedSweptSolid Geometry/Body SectionedSolidHorizontal/DocTemplateDefinition.xml:
```diff
@@ -16,18 +16,7 @@
-														</DocModelRuleConstraint>
-													</Rules>
-												</DocModelRuleEntity>
+												<DocModelRuleEntity Name="IfcLabel" UniqueId="b17762fb-13c9-40ea-9484-491fbcc255ea" />
```
As far as I can tell @larswik the rules should be present in the serialisation of the sub templates, this is the case with the mvdXML encoding also subtemplates redefine the tree and rules of the parent and must match to be valid.

Hence why the rule is serialised on this template also. so please revert 134270f as the previous was the correct serialisation

AlexBrad1eyCT replied:
```suggestion
												<DocModelRuleEntity Name="IfcLabel" href="IfcLabel_2fgI0VXve2tV6eMzYUyfa9"/>
```
this will reference and load the rule correctly in IfcDoc

AlexBrad1eyCT reviewed (changes requested):
just Restore the Body Contraint then it LFTM

AlexBrad1eyCT reviewed (changes requested):
correct reference on body rule

larswik committed:
Update IFC4x3/Templates/Product Shape/Product Geometric Representation/Body Geometry/Body AdvancedSweptSolid Geometry/Body SectionedSolidHorizontal/DocTemplateDefinition.xml

Co-authored-by: Alex Bradley <alex.bradley@abcon.tech>

AlexBrad1eyCT commented on IFC4x3/Templates/Product Shape/Product Geometric Representation/Body Geometry/Body AdvancedSweptSolid Geometry/Body SectionedSolidHorizontal/DocTemplateDefinition.xml:
```diff
@@ -16,18 +16,7 @@
-														</DocModelRuleConstraint>
-													</Rules>
-												</DocModelRuleEntity>
+												<DocModelRuleEntity Name="IfcLabel" href="IfcLabel_2fgI0VXve2tV6eMzYUyfa9"/>
```
```suggestion
												<DocModelRuleEntity Name="IfcLabel" xsi:nil="true" href="IfcLabel_2fgI0VXve2tV6eMzYUyfa9"/>
```

AlexBrad1eyCT committed:
Update IFC4x3/Templates/Product Shape/Product Geometric Representation/Body Geometry/Body AdvancedSweptSolid Geometry/Body SectionedSolidHorizontal/DocTemplateDefinition.xml

AlexBrad1eyCT reviewed (approved):
LFTM!

AlexBrad1eyCT commented:
@jmirtsch @SergejMuhic @pjanck Please provide review as necessary ASAP so we can close this PR

jmirtsch reviewed (commented):
It looks fine to me.  A different but related topic, should there also be a concept template for referenced geometry for sectioned surfaces and solids?  Seems to me this is probably needed for earthwork cuts on road/rail corridors as it's a voiding relationship.

AlexBrad1eyCT commented:
That is a good point @jmirtsch we will have to review and put that together during the MVD work

pjanck commented:
> A different but related topic, should there also be a concept template for referenced geometry for sectioned surfaces and solids? Seems to me this is probably needed for earthwork cuts on road/rail corridors as it's a voiding relationship.

New issue? @AlexBrad1eyCT ?

pjanck commented on IFC4x3/Templates/Product Shape/Product Geometric Representation/Body Geometry/Body AdvancedSweptSolid Geometry/Body SectionedSolidHorizontal/DocTemplateDefinition.xml:
```diff
@@ -16,18 +16,7 @@
-														</DocModelRuleConstraint>
-													</Rules>
-												</DocModelRuleEntity>
+												<DocModelRuleEntity Name="IfcLabel" xsi:nil="true" href="IfcLabel_2fgI0VXve2tV6eMzYUyfa9"/>
```
This seems odd: no `RepresentationIdentifier=Body` for _SectionedSolidHorizontal_, although specified to have one in the documentation?

larswik replied:
?? 
![image](https://user-images.githubusercontent.com/25506404/141955080-5bee2131-d3d4-473d-a10b-031e24ec61a2.png)

AlexBrad1eyCT replied:
@pjanck this a reference to the rule in the parent template so when you update the parent it will hopefully now filter down its a recent feature in IFC doc. As lars shows when you open it, it displays correctly in IfcDoc

pjanck reviewed (commented):
I'm confused.

AlexBrad1eyCT commented:
@pjanck, yes I will create if needed during 

> > A different but related topic, should there also be a concept template for referenced geometry for sectioned surfaces and solids? Seems to me this is probably needed for earthwork cuts on road/rail corridors as it's a voiding relationship.
> 
> New issue? @AlexBrad1eyCT ?

@pjanck yes will, create one, and address during the development of the test case for cuts if needed
